File: js/sense/frame.js
/**
 * Engine-independent tick of fly senses.
 * Field names match fly-brain-minecraft SensoryFrame.java
 * (head frame: azimuth 0 = ahead, positive = fly's right).
 */
function emptyFrame (nColumns) {
  return {
    luminance: Array(nColumns).fill(NaN),
    objects: [],
    yawRateDegPerS: 0,
    pitchRateDegPerS: 0,
    rollRateDegPerS: 0,
    odor: {},
    odorBearingDeg: null,
    taste: {},
    windLeft: 0,
    windRight: 0,
    tilt: 0,
    soundLow: 0,
    soundHigh: 0,
    song: 0,
    touchHead: 0,
    touchWing: 0,
    touchLegs: 0,
    touchNotum: 0,
    touchAbdomen: 0,
    groomDust: 0,
    damage: 0,
    hot: 0,
    cold: 0,
    dry: 0,
    moist: 0,
    airborne: false,
    legsOnGround: true,
    wingbeat: 0,
    objectChannels: { LC4: 0, LPLC2: 0, LC11: 0, LC18: 0, LC10a: 0, LC15: 0, HS: 0, VS: 0 },
  }
}

function addOdor (frame, glom, drive) {
  frame.odor[glom] = (frame.odor[glom] || 0) + drive
}

function addTaste (frame, grn, drive) {
  frame.taste[grn] = Math.max(frame.taste[grn] || 0, drive)
}

function clampOdor (frame, max = 1.5) {
  for (const k of Object.keys(frame.odor)) {
    frame.odor[k] = Math.min(max, frame.odor[k])
  }
}

function hill (s, n = 1.5, k = 0.2) {
  if (s <= 0) return 0
  const sn = s ** n
  return sn / (k ** n + sn)
}

module.exports = { emptyFrame, addOdor, addTaste, clampOdor, hill }

File: python/sense/frame.py
"""Pack a SensoryFrame (fly-brain-minecraft fields) into a fixed vector."""

from __future__ import annotations

import json
from pathlib import Path

GLOMERULI = [
    "DM1", "DM2", "DM3", "DM4", "DM5",
    "VA2", "VA3", "VA6", "VA1v", "VA1d",
    "DL3", "DL4", "DL5",
    "DC2", "DA1", "DA2",
    "VM1", "VM2", "VM4", "VM5d",
    "VL2a", "VC1", "VC2", "VC5",
    "DP1l", "D", "V",
    "VP2", "VP3a", "VP3b", "VP4", "VP5",
]

GRNS = [
    "LB3b", "LB3c", "LB3a", "LB3d",
    "LB1a", "LB1b", "LB1c", "LB1d", "LB1e",
    "PhG1a", "PhG1b", "PhG1c", "PhG3", "PhG4", "PhG16",
    "LgLG3", "LgLG4", "LgAG1", "WG2",
]

OBJECT_CHS = ["LC4", "LPLC2", "LC11", "LC18", "LC10a", "LC15", "HS"]

MECHANO_KEYS = [
    "windLeft", "windRight", "tilt",
    "soundLow", "soundHigh", "song",
    "touchHead", "touchWing", "touchLegs", "touchNotum", "touchAbdomen",
    "groomDust", "damage", "hot", "cold", "dry", "moist",
    "airborne", "legsOnGround", "wingbeat",
]

FEATURE_GROUPS = {
    "glomeruli": GLOMERULI,
    "grns": GRNS,
    "object_channels": OBJECT_CHS,
    "mechano": MECHANO_KEYS,
}


def _retina_n(frame: dict) -> int:
    lum = frame.get("luminance") or []
    return len(lum)


def frame_to_vector(frame: dict, n_retina: int | None = None) -> list[float]:
    lum = list(frame.get("luminance") or [])
    n = n_retina or len(lum) or 64
    if len(lum) < n:
        lum = lum + [0.0] * (n - len(lum))
    vec = [0.0 if v is None or (isinstance(v, float) and v != v) else float(v) for v in lum[:n]]

    odor = frame.get("odor") or {}
    for g in GLOMERULI:
        vec.append(float(odor.get(g, 0.0)))
    bearing = frame.get("odorBearingDeg")
    if bearing is None:
        vec.extend([0.0, 0.0])
    else:
        import math
        rad = float(bearing) * math.pi / 180.0
        vec.extend([math.sin(rad), math.cos(rad)])

    taste = frame.get("taste") or {}
    for g in GRNS:
        vec.append(float(taste.get(g, 0.0)))

    ch = frame.get("objectChannels") or {}
    for k in OBJECT_CHS:
        vec.append(float(ch.get(k, 0.0)))

    for k in MECHANO_KEYS:
        v = frame.get(k, 0)
        vec.append(1.0 if v is True else 0.0 if v is False else float(v or 0))

    vec.append(float(frame.get("yawRateDegPerS") or 0.0) / 180.0)
    vec.append(float(frame.get("pitchRateDegPerS") or 0.0) / 90.0)
    return vec


def vector_size(n_retina: int = 64) -> int:
    return n_retina + len(GLOMERULI) + 2 + len(GRNS) + len(OBJECT_CHS) + len(MECHANO_KEYS) + 2


def load_json(path: str | Path):
    return json.loads(Path(path).read_text())
